module/VuFind/src/VuFind/Db/Service/UserListService.php

```diff
@@ -0,0 +1,415 @@
+     * @param User|int $user    The user posting the tag.
+     * @param UserList $list    The userlist to tag.
+     *
+     * @return void
+     */
+    public function addListTag($tagText, $user, $list)
+    {
+        $tagText = trim($tagText);
+        if (!empty($tagText)) {
+            $tagService = $this->getDbService(\VuFind\Db\Service\TagService::class);
+            $tag = $tagService->getByText($tagText);
+            $tagService->createLink(
+                $tag,
+                null,
+                $user,
+                $list
+            );
+        }
+    }
+
+    /**
+     * Remember that this list was used so that it can become the default in
+     * dialog boxes.
+     *
+     * @param UserList $list User list to be set as default
+     *
+     * @return void
+     */
+    public function rememberLastUsed($list)
+    {
+        if (null !== $this->session) {
+            $this->session->lastUsed = $list->getId();
+        }
+    }
+
+    /**
+     * Get lists containing a specific user_resource
+     *
+     * @param string $resourceId ID of record being checked.
+     * @param string $source     Source of record to look up
+     * @param ?int   $userId     Optional user ID (to limit results to a particular
+     * user).
+     *
+     * @return array
+     */
+    public function getListsContainingResource(
+        string $resourceId,
+        string $source = DEFAULT_SEARCH_BACKEND,
+        ?int $userId = null
+    ): array {
+        $dql = 'SELECT DISTINCT(ul.id), ul FROM ' . $this->getEntityClass(UserList::class) . ' ul '
+            . 'JOIN ' . $this->getEntityClass(UserResource::class) . ' ur WITH ur.list = ul.id '
+            . 'JOIN ' . $this->getEntityClass(Resource::class) . ' r WITH r.id = ur.resource '
+            . 'WHERE r.recordId = :resourceId AND r.source = :source ';
+
+        $parameters = compact('resourceId', 'source');
+        if (null !== $userId) {
+            $dql .= 'AND ur.user = :userId ';
+            $parameters['userId'] = $userId;
+        }
+
+        $dql .= 'ORDER BY ul.title';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $results = $query->getResult();
+        return $results;
+    }
+
+    /**
+     * Get public lists other than the one in array.
+     *
+     * @param array $lists List of public lists.
+     *
+     * @return array
+     */
+    public function getPublicLists($lists)
+    {
+        $dql = 'SELECT ul FROM ' . $this->getEntityClass(UserList::class) . ' ul '
+            . 'WHERE ul NOT IN (:lists) AND ul.public = 1 ';
+        $parameters = compact('lists');
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $results = $query->getResult();
+        return $results;
+    }
+
+    /**
+     * Get all of the lists associated with this user.
+     *
+     * @param User|int $user Id of the user owning the list.
```

Comment:
It appears that we're currently only calling this with an integer. Are we confident that it will work with a User too? The param comment specifically says "ID of the user..." so we should either adjust the type or adjust the description. Whatever we change here also needs to be adjusted in the UserList view helper.

Reply:
This will work with the user too. Once we migrate the User some the calls may pass in a User object.

Reply:
In that case, we should just edit the comment to say "User object or ID representing the user owning the list."